### DISCUSSION TITLE
How to properly bundle Web Workers?

### DISCUSSION THREAD

**Discussion by zimmed** [oven-sh/bun#14069] [Q&A] [answered]
I've been at this for a little while now, and I can't seem to figure out how to get Bun to handle web worker files when creating a javascript bundle. So far, the only thing I've gotten to work, is to manually build workers as entries, and use the final built path of the worker file in my source code so once it's transpiled, it references the file correctly. This is far from ideal, obviously. I should be able to reference a relative-path typescript file in my source code, and have it replaced with the relative path to the transpiled asset at build time.

For compiling single executables, this appears to already be handled (https://github.com/oven-sh/bun/blob/main/docs/bundler/executables.md#worker) but as far as I can tell, no such feature exists when bundling?

Is there something simple I'm missing, or does this functionality not yet exist?

**Comment by zimmed** [comment, marked as answer]
I ended up hacking together build plugins to handle implicit building of workers and rebundling asset files, which works surprisingly well.

Would still be great if Bun handled this automatically, but it's definitely not a trivial operation.

**Reply by eighty4** [reply]
could you share a code sample?

**Reply by JordanMajd** [reply]
Not OP but my solution to bundle a web worker is two parts, for development server and for building production.

If you are registering a web worker like so: 

`my-web-page.ts`
```
const registration = await navigator.serviceWorker.register(
  "/my-service-worker.js",
  {
    type: "module",
    scope: "/",
  },
);
```

1. For the development server you can create a route that transpiles and serves the js file: 

`server.ts`
```
const server = serve({
  routes: {
     // make sure this route matches the one you are registering
    "/my-service-worker.js": async () => {
      const source = await Bun.file("./src/workers/my-service-worker.ts").text();
      const code = transpiler.transformSync(source);
      return new Response(code, {
        headers: {
          "Content-Type": "application/javascript; charset=utf-8",
          "Cache-Control": "no-store",
        },
      });
    },
    "/*": index,
  },
  development: process.env.NODE_ENV !== "production" && {
    hmr: true,
    console: true,
  },
});
```

2. To get it to work when building add the webworker file as a target to your build command. In `package.json` add `./src/workers/my-service-worker.ts` as a target to the build command and also preserve the naming with `entry-naming='[name].[ext]'`

`package.json`
```
"scripts": {
   "build": "bun build ./src/index.html ./src/workers/my-service-worker.ts --entry-naming='[name].[ext]' --outdir=dist --sourcemap --target=browser --minify --define:process.env.NODE_ENV='\"production\"' --env='BUN_PUBLIC_*'",
 }
 ```

**Reply by joeriexelmans** [reply]
@JordanMajd Thanks for your reply. I will try to do something based on your example, but your examples are about **service workers,** which serve a different purpose from **web workers**. But I guess in both cases you want to compile TS code to a separate JS file.
